pennylane/qnode_new/qnode.py

```diff
@@ -0,0 +1,935 @@
+
+        return tuple(pos), kwargs
+
+
+    def _default_args(self, kwargs):
+        """Validate the quantum function arguments, apply defaults.
+
+        Here we apply default values for the non-differentiable parameters of :attr:`QNode.func`.
+
+        Args:
+            kwargs (dict[str, Any]): auxiliary arguments (given using the keyword syntax)
+
+        Returns:
+            dict[str, Any]: all auxiliary arguments (with defaults)
+        """
+        forbidden_kinds = (inspect.Parameter.POSITIONAL_ONLY,
+                           inspect.Parameter.VAR_POSITIONAL,
+                           inspect.Parameter.VAR_KEYWORD)
+
+        # check the validity of kwargs items
+        for name in kwargs:
+            s = self.func.sig.get(name)
+            if s is None:
+                if self.func.var_keyword:
+                    continue  # unknown parameter, but **kwargs will take it TODO should it?
+                raise QuantumFunctionError("Unknown quantum function parameter '{}'.".format(name))
+            if s.par.kind in forbidden_kinds or s.par.default == inspect.Parameter.empty:
+                raise QuantumFunctionError("Quantum function parameter '{}' cannot be given using the keyword syntax.".format(name))
+
+        # apply defaults
+        for name, s in self.func.sig.items():
+            default = s.par.default
+            if default != inspect.Parameter.empty:
+                # meant to be given using keyword syntax
+                kwargs.setdefault(name, default)
+
+        return kwargs
+
+
+    def __call__(self, *args, **kwargs):
+        """Wrapper for :meth:`~.QNode.evaluate`.
+        """
+        #args, kwargs = self._sort_args(args, kwargs)
+        return self.evaluate(args, kwargs)
+
+
+    def evaluate(self, args, kwargs):
+        """Evaluate the quantum function on the specified device.
+
+        Args:
+            args (tuple[Any]): positional arguments to the quantum function (differentiable)
+            kwargs (dict[str, Any]): auxiliary arguments (not differentiable)
+
+        Returns:
+            float or array[float]: output measured value(s)
+        """
+        #args, kwargs = self._check_args(args, kwargs)
+        kwargs = self._default_args(kwargs)
+
+        if self.circuit is None or self.mutable:
+            if self.circuit is not None:
+                # circuit construction has previously been called
+                flat_args = list(_flatten(args))
+                if len(flat_args) == self.num_variables:
+                    # only construct the circuit if the number
+                    # of arguments matches the allowed number
+                    # of variables.
+                    # This avoids construction happening
+                    # via self._pd_analytic, where temporary
+                    # variables are appended to the positional argument list.
+
+                    # FIXME maybe we require that if extra args are given, the caller must _construct()
```

Comment:
This part here is waiting for the next step, JacobianQNode.

Reply:
Note: I wrote the above, but I don't think it is entirely correct and probably buggy. There should be a better way of working out if a gradient method if calling evaluate than heuristics looking at the number of arguments

Reply:
Yeah, I think when `evaluate` is called during Jacobian calculation you never need to call `_construct` since `JacobianQNode.jacobian` has already called it once, and only the positional args are varied during Jacobian calculation.

Reply:
I made it so that `JacobianQNode.jacobian` (coming in the next PR) explicitly changes the QNode to immutable after calling `QNode._construct`, and back to the original mutability when it's done. This way the check in `QNode.evaluate` becomes really simple.